pennylane/qinfo/classical_fisher.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Differentiable classical fisher information"""
+# pylint: disable=import-outside-toplevel
+import functools
+import pennylane as qml
+
+from pennylane.transforms import batch_transform
+
+
+def _torch_jac(circ):
+    """Torch jacobian as a callable function"""
+    import torch
+
+    def wrapper(*args, **kwargs):
+        loss = functools.partial(circ, **kwargs)
+        if len(args) > 1:
+            return torch.autograd.functional.jacobian(loss, (args))
+        return torch.autograd.functional.jacobian(loss, *args)
+
+    return wrapper
+
+
+def _tf_jac(circ):
+    """TF jacobian as a callable function"""
+    import tensorflow as tf
+
+    def wrapper(*args, **kwargs):
+        with tf.GradientTape() as tape:
+            loss = circ(*args, **kwargs)
+        return tape.jacobian(loss, (args))
+
+    return wrapper
+
+
+def CFIM(qnode, argnums=0):
+    """Computing the classical fisher information matrix (CFIM) using the jacobian of the output probabilities
+    as described in eq. (15) in https://arxiv.org/abs/2103.15191
+    """
+    new_qnode = _make_probs(qnode, post_processing_fn=lambda x: qml.math.squeeze(qml.math.stack(x)))
+
+    interface = qnode.interface
+
+    if interface == "jax":
+        import jax
+
+        jac = jax.jacobian(new_qnode, argnums=argnums)
+
+    if interface == "torch":
+        jac = _torch_jac(new_qnode)
+
+    if interface == "autograd":
+        jac = qml.jacobian(new_qnode)
+
+    if interface == "tf":
+        jac = _tf_jac(new_qnode)
+
+    def wrapper(*args, **kwargs):
+        p = qnode(*args, **kwargs)
+        j = jac(*args, **kwargs)
+
+        # In case multiple variables are used
+        if isinstance(j, tuple):
+            return [_compute_cfim(p, j_i, interface) for j_i in j]
+
+        return _compute_cfim(p, j, interface)
+
+    return wrapper
+
+
+def _compute_cfim(p, dp, interface):
+    r"""Computes the (num_params, num_params) classical fisher information matrix from the probabilities and its derivatives
+    I.e. it computes :math:`CFIM_{ij} = \sum_\ell (\partial_i p_\ell) (\partial_i p_\ell) / p_\ell`
+    """
+
+    # Compute 1/p for p!=0, else 0
+    if any(qml.math.isclose(p, qml.math.zeros_like(p))) and not interface == "tf":
```

Comment:
Could remove the zeros instead of removing the `Nan` values? We add similar problem for `log(0)` in entropy and we used `qml.math.where` to replace them.

Reply:
I am not sure what you mean by that, removing the `Nan` values?
My first attempt was to use `qml.math.where`, look for the zeros, and then only process the non-zeros leaving the rest 0s as 0s. But I had problems with some interfaces, thats why this function is super messy at the moment. I agree that creating a function for that in the math module is a very good idea!